File: JavaTest/src/main/java/test/ArrayListTest.java
package test;

import java.util.ArrayList;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * The {@code ArrayListTest} class performs automatic testing of the
 * {@code ArrayList} class.
 *
 * @author Petr Kozler
 */
public class ArrayListTest {

    public static Logger logger = LogManager.getLogger(ArrayListTest.class.getName());

    public static void runTestSuite() {
        ArrayListTest.arrayListTest();
        ArrayListTest.arrayListTest2();
        ArrayListTest.addTest();
        ArrayListTest.addTest2();
        ArrayListTest.getTest();
        ArrayListTest.setTest();
        ArrayListTest.removeTest();
        ArrayListTest.sizeTest();
        ArrayListTest.isEmptyTest();
        ArrayListTest.clearTest();
        ArrayListTest.toStringTest();
    }

    /**
     * Tests the constructor.
     */
    public static void arrayListTest() {
        ArrayList<TestObject> a;
        String b;
        // empty list
        a = new ArrayList<>();
        b = a.toString();
        logger.info(b);
    }

    /**
     * Tests the constructor with ArrayList as a parameter.
     */
    public static void arrayListTest2() {
        ArrayList<TestObject> a, b;
        TestObject c;
        String d;
        // creating the list from an empty list
        a = new ArrayList<>();
        b = new ArrayList<>(a);
        d = b.toString();
        logger.info(d);
        // creating the list from a list with one element
        a = new ArrayList<>();
        c = new TestObject("first", 1);
        a.add(c);
        b = new ArrayList<>(a);
        d = b.toString();
        logger.info(d);
        // creating the list from a list with some elements
        a = new ArrayList<>();
        c = new TestObject("first", 1);
        a.add(c);
        c = new TestObject("second", 2);
        a.add(c);
        c = new TestObject("third", 3);
        a.add(c);
        b = new ArrayList<>(a);
        d = b.toString();
        logger.info(d);
    }

    /**
     * Tests the add method.
     */
    public static void addTest() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        // adding some elements at the end of the list
        a = new ArrayList<>();
        b = new TestObject("first", 1);
        a.add(b);
        c = a.toString();
        logger.info(c);
        b = new TestObject("second", 2);
        a.add(b);
        c = a.toString();
        logger.info(c);
        b = new TestObject("first", 2);
        a.add(b);
        c = a.toString();
        logger.info(c);
        b = new TestObject("second", 1);
        a.add(b);
        c = a.toString();
        logger.info(c);
        b = new TestObject("first", 1);
        a.add(b);
        c = a.toString();
        logger.info(c);
    }

    /**
     * Tests the add method with index as a parameter.
     */
    public static void addTest2() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        a = new ArrayList<>();
        // adding an element at the beginning
        b = new TestObject("first", 1);
        a.add(0, b);
        c = a.toString();
        logger.info(c);
        // adding an element at the end
        b = new TestObject("second", 2);
        a.add(1, b);
        c = a.toString();
        logger.info(c);
        // adding an element at the beginning
        b = new TestObject("third", 3);
        a.add(0, b);
        c = a.toString();
        logger.info(c);
        // adding an element in the middle
        b = new TestObject("fourth", 4);
        a.add(1, b);
        c = a.toString();
        logger.info(c);
        // adding an element in the middle
        b = new TestObject("fifth", 5);
        a.add(2, b);
        c = a.toString();
        logger.info(c);
    }

    /**
     * Tests the get method.
     */
    public static void getTest() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        a = new ArrayList<>();
        // getting an element from the beginning
        b = new TestObject("first", 1);
        a.add(b);
        b = a.get(0);
        c = b.toString();
        logger.info(c);
        // getting an element from the beginning (more elements in the list)
        b = new TestObject("second", 2);
        a.add(b);
        b = a.get(0);
        c = b.toString();
        logger.info(c);
        // getting an element from the end
        b = a.get(1);
        c = b.toString();
        logger.info(c);
        b = new TestObject("third", 3);
        a.add(b);
        b = new TestObject("fourth", 4);
        a.add(b);
        b = new TestObject("fifth", 5);
        a.add(b);
        // getting an element from the beginning
        b = a.get(0);
        c = b.toString();
        logger.info(c);
        // getting an element from the end
        b = a.get(4);
        c = b.toString();
        logger.info(c);
    }

    /**
     * Tests the set method.
     */
    public static void setTest() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        a = new ArrayList<>();
        b = new TestObject("first", 1);
        a.add(b);
        b = new TestObject("second", 2);
        a.add(b);
        b = new TestObject("third", 3);
        a.add(b);
        b = new TestObject("fourth", 4);
        a.add(b);
        b = new TestObject("fifth", 5);
        a.add(b);
        b = new TestObject("sixth", 6);
        // replacing an element at the beginning
        a.set(0, b);
        c = a.toString();
        logger.info(c);
        // replacing an element in the middle
        b = new TestObject("seventh", 7);
        a.set(2, b);
        c = a.toString();
        logger.info(c);
        // replacing an element at the end
        b = new TestObject("eighth", 8);
        a.set(4, b);
        c = a.toString();
        logger.info(c);
    }

    /**
     * Tests the remove method.
     */
    public static void removeTest() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        a = new ArrayList<>();
        b = new TestObject("first", 1);
        a.add(b);
        b = new TestObject("second", 2);
        a.add(b);
        b = new TestObject("third", 3);
        a.add(b);
        b = new TestObject("fourth", 4);
        a.add(b);
        b = new TestObject("fifth", 5);
        // removing an element from the beginning
        a.add(b);
        a.remove(0);
        c = a.toString();
        logger.info(c);
        // removing the element from the middle
        a.remove(1);
        c = a.toString();
        logger.info(c);
        // removing an element from the end
        a.remove(2);
        c = a.toString();
        logger.info(c);
    }

    /**
     * Tests the size method.
     */
    public static void sizeTest() {
        ArrayList<TestObject> a;
        TestObject b;
        int c;
        a = new ArrayList<>();
        // size in the beginning
        c = a.size();
        logger.info(c);
        b = new TestObject("first", 1);
        // size after adding one element
        a.add(b);
        c = a.size();
        logger.info(c);
        // size after adding more elements
        b = new TestObject("second", 2);
        a.add(b);
        b = new TestObject("third", 3);
        a.add(b);
        c = a.size();
        logger.info(c);
        // size after removing one element
        a.remove(0);
        c = a.size();
        logger.info(c);
        // size after removing more elements
        a.remove(0);
        a.remove(0);
        c = a.size();
        logger.info(c);
    }

    /**
     * Tests the isEmpty method.
     */
    public static void isEmptyTest() {
        ArrayList<TestObject> a;
        TestObject b;
        boolean c;
        a = new ArrayList<>();
        c = a.isEmpty();
        logger.info(c);
        b = new TestObject("first", 1);
        a.add(b);
        c = a.isEmpty();
        logger.info(c);
        b = new TestObject("second", 2);
        a.add(b);
        b = new TestObject("third", 3);
        a.add(b);
        c = a.isEmpty();
        logger.info(c);
        a.remove(0);
        c = a.isEmpty();
        logger.info(c);
        a.remove(0);
        a.remove(0);
        c = a.isEmpty();
        logger.info(c);
    }

    /**
     * Tests the clear method.
     */
    public static void clearTest() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        a = new ArrayList<>();
        // clear in the beginning
        a.clear();
        c = a.toString();
        logger.info(c);
        // clear after adding one element
        b = new TestObject("first", 1);
        a.add(b);
        a.clear();
        c = a.toString();
        logger.info(c);
        // clear after adding more elements
        b = new TestObject("first", 1);
        a.add(b);
        b = new TestObject("second", 2);
        a.add(b);
        b = new TestObject("third", 3);
        a.add(b);
        a.clear();
        c = a.toString();
        logger.info(c);
    }

    /**
     * Tests the toString method.
     */
    public static void toStringTest() {
        ArrayList<TestObject> a;
        TestObject b;
        String c;
        // the text representation of an empty list
        a = new ArrayList<>();
        c = a.toString();
        logger.info(c);
        // the text representation of a list with one element
        b = new TestObject("first", 1);
        a.add(b);
        c = a.toString();
        logger.info(c);
        // the text representation of a list with some elements
        b = new TestObject("second", 2);
        a.add(b);
        b = new TestObject("third", 3);
        a.add(b);
        c = a.toString();
        logger.info(c);
    }

}

File: C/CLibrary/tests/ArrayListTest.c
#include <stdio.h>
#include <stdlib.h>

#include "test.h"
#include "ArrayList.h"
#include "TestObject.h"

/**
 * An ArrayList module test suite.
 *
 * @author Petr Kozler (A13B0359P)
 */

/*
    Creates the array list from array.
 */
ArrayList *toList(TestObject **array, int32_t length) {
    ArrayList *list = new_ArrayList();

    int32_t i;
    for (i = 0; i < length; i++) {
        addAl(list, array[i]);
    }
    
    return list;
}

/**
 * Tests the constructor.
 */

void arrayListTest() {
    const int32_t length = 0;
    ArrayList *a;
    bool b;
    TestObject **c;
    // empty list
    a = new_ArrayList();
    c = malloc(length * sizeof(TestObject *));
    b = equalsList(a, sizeAl, getAl, toList(c, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, b);
}

/**
 * Tests the constructor with ArrayList as a parameter.
 */

void arrayListTest2() {
    const int32_t length = 0;
    const int32_t length2 = 1;
    const int32_t length3 = 3;
    ArrayList *a, *b;
    TestObject *c;
    bool d;
    TestObject **e;
    // creating the list from an empty list
    a = new_ArrayList();
    b = new_ArrayListAddAll(a);
    e = malloc(length * sizeof(TestObject *));
    d = equalsList(b, sizeAl, getAl, toList(e, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, d);
    // creating the list from a list with one element
    a = new_ArrayList();
    c = new_TestObject(new_String("first"), 1);
    addAl(a, c);
    b = new_ArrayListAddAll(a);
    e = malloc(length2 * sizeof(TestObject *));
    e[0] = new_TestObject(new_String("first"), 1);
    d = equalsList(b, sizeAl, getAl, toList(e, length2), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, d);
    // creating the list from a list with some elements
    a = new_ArrayList();
    e = malloc(length3 * sizeof(TestObject *));
    e[0] = new_TestObject(new_String("first"), 1);
    e[1] = new_TestObject(new_String("second"), 2);
    e[2] = new_TestObject(new_String("third"), 3);
    c = new_TestObject(new_String("first"), 1);
    addAl(a, c);
    c = new_TestObject(new_String("second"), 2);
    addAl(a, c);
    c = new_TestObject(new_String("third"), 3);
    addAl(a, c);
    b = new_ArrayListAddAll(a);
    d = equalsList(b, sizeAl, getAl, toList(e, length3), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, d);
}

/**
 * Tests the addAl function.
 */

void addTest() {
    const int32_t length = 1;
    const int32_t length2 = 2;
    const int32_t length3 = 3;
    const int32_t length4 = 4;
    const int32_t length5 = 5;
    ArrayList *a;
    TestObject **d;
    TestObject *b;
    bool c;
    // adding some elements at the end of the list
    a = new_ArrayList();
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    d = malloc(length2 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    d[1] = new_TestObject(new_String("second"), 2);
    c = equalsList(a, sizeAl, getAl, toList(d, length2), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    b = new_TestObject(new_String("first"), 2);
    addAl(a, b);
    d = malloc(length3 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("first"), 2);
    c = equalsList(a, sizeAl, getAl, toList(d, length3), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    b = new_TestObject(new_String("second"), 1);
    addAl(a, b);
    d = malloc(length4 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("first"), 2);
    d[3] = new_TestObject(new_String("second"), 1);
    c = equalsList(a, sizeAl, getAl, toList(d, length4), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    d = malloc(length5 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("first"), 2);
    d[3] = new_TestObject(new_String("second"), 1);
    d[4] = new_TestObject(new_String("first"), 1);
    c = equalsList(a, sizeAl, getAl, toList(d, length5), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
}

/**
 * Tests the addAl function with index as a parameter.
 */

void addTest2() {
    const int32_t length = 1;
    const int32_t length2 = 2;
    const int32_t length3 = 3;
    const int32_t length4 = 4;
    const int32_t length5 = 5;
    ArrayList *a;
    TestObject *b;
    bool c;
    TestObject **d;
    a = new_ArrayList();
    // adding an element at the beginning
    b = new_TestObject(new_String("first"), 1);
    addAtAl(a, 0, b);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // adding an element at the end
    b = new_TestObject(new_String("second"), 2);
    addAtAl(a, 1, b);
    d = malloc(length2 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 1);
    d[1] = new_TestObject(new_String("second"), 2);
    c = equalsList(a, sizeAl, getAl, toList(d, length2), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // adding an element at the beginning
    b = new_TestObject(new_String("third"), 3);
    addAtAl(a, 0, b);
    d = malloc(length3 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("third"), 3);
    d[1] = new_TestObject(new_String("first"), 1);
    d[2] = new_TestObject(new_String("second"), 2);
    c = equalsList(a, sizeAl, getAl, toList(d, length3), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // adding an element in the middle
    b = new_TestObject(new_String("fourth"), 4);
    addAtAl(a, 1, b);
    d = malloc(length4 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("third"), 3);
    d[1] = new_TestObject(new_String("fourth"), 4);
    d[2] = new_TestObject(new_String("first"), 1);
    d[3] = new_TestObject(new_String("second"), 2);
    c = equalsList(a, sizeAl, getAl, toList(d, length4), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // adding an element in the middle
    b = new_TestObject(new_String("fifth"), 5);
    addAtAl(a, 2, b);
    d = malloc(length5 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("third"), 3);
    d[1] = new_TestObject(new_String("fourth"), 4);
    d[2] = new_TestObject(new_String("fifth"), 5);
    d[3] = new_TestObject(new_String("first"), 1);
    d[4] = new_TestObject(new_String("second"), 2);
    c = equalsList(a, sizeAl, getAl, toList(d, length5), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
}

/**
 * Tests the equalsList function.
 */

void equalsTest() {
    const int32_t length = 5;
    const int32_t length2 = 1;
    ArrayList *a, *b;
    bool c;
    TestObject **d;
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 3);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("third"), 8);
    d[3] = new_TestObject(new_String("fourth"), 7);
    d[4] = new_TestObject(new_String("fifth"), 6);
    a = toList(d, length);
    b = NULL;
    c = equalsList(a, sizeAl, getAl, b, sizeAl, getAl, equalsTo);
    assertEqualsBool(false, c);
    d = malloc(length2 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 3);
    b = toList(d, length2);
    c = equalsList(a, sizeAl, getAl, b, sizeAl, getAl, equalsTo);
    assertEqualsBool(false, c);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("first"), 3);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("third"), 8);
    d[3] = new_TestObject(new_String("fourth"), 7);
    d[4] = new_TestObject(new_String("fifth"), 3);
    b = toList(d, length);
    c = equalsList(a, sizeAl, getAl, b, sizeAl, getAl, equalsTo);
    assertEqualsBool(false, c);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("sixth"), 3);
    d[1] = new_TestObject(new_String("seventh"), 2);
    d[2] = new_TestObject(new_String("eighth"), 8);
    d[3] = new_TestObject(new_String("nineth"), 7);
    d[4] = new_TestObject(new_String("tenth"), 6);
    b = toList(d, length);
    c = equalsList(a, sizeAl, getAl, b, sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    b = a;
    c = equalsList(a, sizeAl, getAl, b, sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
}

/**
 * Tests the getAl function.
 */

void getTest() {
    ArrayList *a;
    TestObject *b;
    bool c;
    a = new_ArrayList();
    // getting an element from the beginning
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    b = getAl(a, 0);
    c = equalsTo(b, new_TestObject(new_String("first"), 1));
    assertEqualsBool(true, c);
    // getting an element from the beginning (more elements in the list)
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = getAl(a, 0);
    c = equalsTo(b, new_TestObject(new_String("first"), 1));
    assertEqualsBool(true, c);
    // getting an element from the end
    b = getAl(a, 1);
    c = equalsTo(b, new_TestObject(new_String("second"), 2));
    assertEqualsBool(true, c);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    b = new_TestObject(new_String("fourth"), 4);
    addAl(a, b);
    b = new_TestObject(new_String("fifth"), 5);
    addAl(a, b);
    // getting an element from the beginning
    b = getAl(a, 0);
    c = equalsTo(b, new_TestObject(new_String("first"), 1));
    assertEqualsBool(true, c);
    // getting an element from the end
    b = getAl(a, 4);
    c = equalsTo(b, new_TestObject(new_String("fifth"), 5));
    assertEqualsBool(true, c);
}

/**
 * Tests the setAl function.
 */

void setTest() {
    const int32_t length = 5;
    ArrayList *a;
    TestObject *b;
    bool c;
    TestObject **d;
    a = new_ArrayList();
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    b = new_TestObject(new_String("fourth"), 4);
    addAl(a, b);
    b = new_TestObject(new_String("fifth"), 5);
    addAl(a, b);
    b = new_TestObject(new_String("sixth"), 6);
    // replacing an element at the beginning
    setAl(a, 0, b);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("sixth"), 6);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("third"), 3);
    d[3] = new_TestObject(new_String("fourth"), 4);
    d[4] = new_TestObject(new_String("fifth"), 5);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // replacing an element in the middle
    b = new_TestObject(new_String("seventh"), 7);
    setAl(a, 2, b);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("sixth"), 6);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("seventh"), 7);
    d[3] = new_TestObject(new_String("fourth"), 4);
    d[4] = new_TestObject(new_String("fifth"), 5);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // replacing an element at the end
    b = new_TestObject(new_String("eighth"), 8);
    setAl(a, 4, b);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("sixth"), 6);
    d[1] = new_TestObject(new_String("second"), 2);
    d[2] = new_TestObject(new_String("seventh"), 7);
    d[3] = new_TestObject(new_String("fourth"), 4);
    d[4] = new_TestObject(new_String("eighth"), 8);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
}

/**
 * Tests the remove function.
 */

void removeTest() {
    const int32_t length = 4;
    const int32_t length2 = 3;
    const int32_t length3 = 2;
    ArrayList *a;
    TestObject *b;
    bool c;
    TestObject **d;
    a = new_ArrayList();
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    b = new_TestObject(new_String("fourth"), 4);
    addAl(a, b);
    b = new_TestObject(new_String("fifth"), 5);
    // removing an element from the beginning
    addAl(a, b);
    removeAl(a, 0);
    d = malloc(length * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("second"), 2);
    d[1] = new_TestObject(new_String("third"), 3);
    d[2] = new_TestObject(new_String("fourth"), 4);
    d[3] = new_TestObject(new_String("fifth"), 5);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // removing the element from the middle
    removeAl(a, 1);
    d = malloc(length2 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("second"), 2);
    d[1] = new_TestObject(new_String("fourth"), 4);
    d[2] = new_TestObject(new_String("fifth"), 5);
    c = equalsList(a, sizeAl, getAl, toList(d, length2), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // removing an element from the end
    removeAl(a, 2);
    d = malloc(length3 * sizeof(TestObject *));
    d[0] = new_TestObject(new_String("second"), 2);
    d[1] = new_TestObject(new_String("fourth"), 4);
    c = equalsList(a, sizeAl, getAl, toList(d, length3), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
}

/**
 * Tests the size function.
 */

void sizeTest() {
    ArrayList *a;
    TestObject *b;
    int32_t c;
    a = new_ArrayList();
    // size in the beginning
    c = sizeAl(a);
    assertEqualsI(0, c);
    b = new_TestObject(new_String("first"), 1);
    // size after adding one element
    addAl(a, b);
    c = sizeAl(a);
    assertEqualsI(1, c);
    // size after adding more elements
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    c = sizeAl(a);
    assertEqualsI(3, c);
    // size after removing one element
    removeAl(a, 0);
    c = sizeAl(a);
    assertEqualsI(2, c);
    // size after removing more elements
    removeAl(a, 0);
    removeAl(a, 0);
    c = sizeAl(a);
    assertEqualsI(0, c);
}

/**
 * Tests the isEmpty function.
 */

void isEmptyTest() {
    ArrayList *a;
    TestObject *b;
    bool c;
    a = new_ArrayList();
    c = isEmptyAl(a);
    assertEqualsBool(true, c);
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    c = isEmptyAl(a);
    assertEqualsBool(false, c);
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    c = isEmptyAl(a);
    assertEqualsBool(false, c);
    removeAl(a, 0);
    c = isEmptyAl(a);
    assertEqualsBool(false, c);
    removeAl(a, 0);
    removeAl(a, 0);
    c = isEmptyAl(a);
    assertEqualsBool(true, c);
}

/**
 * Tests the clear function.
 */

void clearTest() {
    const int32_t length = 0;
    ArrayList *a;
    TestObject *b;
    bool c;
    TestObject **d;
    d = malloc(length * sizeof(TestObject *));
    a = new_ArrayList();
    // clear in the beginning
    clearAl(a);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // clear after adding one element
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    clearAl(a);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
    // clear after adding more elements
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    clearAl(a);
    c = equalsList(a, sizeAl, getAl, toList(d, length), sizeAl, getAl, equalsTo);
    assertEqualsBool(true, c);
}

/**
 * Tests the toString function.
 */

void toStringTest() {
    ArrayList *a;
    TestObject *b;
    String *c;
    // the text representation of an empty list
    a = new_ArrayList();
    c = toStringAl(a, toStringTo);
    assertEqualsStr("[]", c->s);
    // the text representation of a list with one element
    b = new_TestObject(new_String("first"), 1);
    addAl(a, b);
    c = toStringAl(a, toStringTo);
    assertEqualsStr("[TestObject [str=first, i=1]]", c->s);
    // the text representation of a list with some elements
    b = new_TestObject(new_String("second"), 2);
    addAl(a, b);
    b = new_TestObject(new_String("third"), 3);
    addAl(a, b);
    c = toStringAl(a, toStringTo);
    assertEqualsStr("[TestObject [str=first, i=1], TestObject [str=second, i=2], TestObject [str=third, i=3]]", c->s);
}

int main(int argc, char** argv) {
    const int testCount = 12;
    setUpTestModule("ArrayListTest", testCount);

    registerTest(arrayListTest, "arrayListTest");
    registerTest(arrayListTest2, "arrayListTest2");
    registerTest(addTest, "addTest");
    registerTest(addTest2, "addTest2");
    registerTest(equalsTest, "equalsTest");
    registerTest(getTest, "getTest");
    registerTest(setTest, "setTest");
    registerTest(removeTest, "removeTest");
    registerTest(sizeTest, "sizeTest");
    registerTest(isEmptyTest, "isEmptyTest");
    registerTest(clearTest, "clearTest");
    registerTest(toStringTest, "toStringTest");
    
    runTests();

    return (EXIT_SUCCESS);
}
